### PR TITLE
Behat ArgumentResolver for Context service injection 

### PR DESCRIPTION
**Pull request by miguelcleverti**
Replaces #44
Resolves arguments using annotations and let Symfony2BehatExtension handle services requirements.

### CODE REVIEW

**Comment by miguelcleverti** [comment]
@bdunogier check this out!

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function resolveArguments(ReflectionClass $classReflection, array $arguments = [])
+    {
+        $injArguments = $this->getGetMethodServices($classReflection);
+
+        if (!empty($injArguments)) {
+            $arguments = [];
+            foreach ($injArguments as $name => $service) {
+                $name = preg_replace('/^\$/', '', $name);
+                var_dump($name);
```
Comment (joaoinacio) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: remove?

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+        preg_match_all('#@(.*?)\n#s', $refMethod->getDocComment(), $matches);
+        // parse array from (numeric key => 'annotation <value>') to (annotation => value)
+        $methodServices = [];
+        foreach ($matches[1] as $annotation) {
+            $split = preg_split('/[\s]+/', $annotation, 3);
+            $tag = $split[0];
+
+            if ($tag == self::SERVICE_DOC_TAG) {
+                $param = isset($split[1]) ? $split[1] : '';
```
Comment (joaoinacio) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: IMO argument and service should be required for this tag, 
Either only use it if both are set, or even throw an exception if either one is missing?

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace EzSystems\BehatBundle\ServiceContainer\Argument;
+
+use ReflectionClass;
+use Behat\Behat\Context\Argument\ArgumentResolver;
+
+/**
+ * Behat Context Argument Resolver
+ */
+class EzBehatArgumentResolver implements ArgumentResolver
```
Comment (bdunogier) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: Should be named after what it does (`AnnotationArgumentResolver` ?).

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace EzSystems\BehatBundle\ServiceContainer\Argument;
```
Comment (bdunogier) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: I'd prefer if we followed the naming from the sf2 ext, and put it in `Context/Resolver`.

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     * Resolve service arguments for Behat Context constructor thru annotation.
+     * Symfony2Extension ArgumentResoler will convert service names to actual instances.
+     *
+     * @param Context $context
+     */
+    public function resolveArguments(ReflectionClass $classReflection, array $arguments = [])
+    {
+        $injArguments = $this->getGetMethodServices($classReflection);
+
+        if (!empty($injArguments)) {
```
Comment (bdunogier) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: `foreach ($this->getGetMethodServices($classReflection)) {`

File: ServiceContainer/Argument/EzBehatArgumentResolver.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        return $arguments;
+    }
+
+    /**
+     * Returns an array with the method arguments service requirements,
+     * if the methods use the service Annotation
+     *
+     * @return array array of methods and their service dependencies
+     */
+    private function getGetMethodServices($refClass, $method = '__construct')
```
Comment (bdunogier) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: `parseAnnotations()` ?

File: ServiceContainer/Argument/EzBehatArgumentResolver.php (hunk 5 above)
Comment (miguelcleverti) [inline ServiceContainer/Argument/EzBehatArgumentResolver.php]: We only what to set the arguments if there are annotations, if not return the default ones.

**Comment by miguelcleverti** [comment]
ping @bdunogier

**Comment by joaoinacio** [comment]
Looks good for merge after squashing, test fail is "expected".

File: Context/Resolver/AnnotationArgumentResolver.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+     * @param Context $context
+     */
+    public function resolveArguments(ReflectionClass $classReflection, array $arguments = [])
+    {
+        $injArguments = $this->parseAnnotations($classReflection);
+
+        if (!empty($injArguments)) {
+            $arguments = [];
+            foreach ($injArguments as $name => $service) {
+                $name = preg_replace('/^\$/', '', $name);
```
Comment (bdunogier) [inline Context/Resolver/AnnotationArgumentResolver.php]: I'd actually take care of this in the private method. Add the `$` to the pcre below, and match the variable name without it

**Comment by bdunogier** [comment]
+1 with the two preg match merged into one.

**Comment by miguelcleverti** [comment]
@bdunogier the annotation fetch now is done in a different method than the parsing, and the preg match matches the exact pattern that we want.

**Comment by bdunogier** [comment]
> the annotation fetch now is done in a different method than the parsing, and the preg match matches the exact pattern that we want.

That wasn't really necessary. We could have grepped `@injectService $([a-z0-9])+ @([a-z0-9-_])` in one unique call, and use that to build the result. But it's not that important, let's merge this.

**Comment by joaoinacio** [comment]
thanks, merging.

**Comment by miguelcleverti** [comment]
@joaoinacio all squashes ready, merge is a go!

**Pull request merged**